centreon/src/Core/Common/Infrastructure/Repository/DataStorageObserver.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * Copyright 2005 - 2023 Centreon (https://www.centreon.com/)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types = 1);
+
+namespace Core\Common\Infrastructure\Repository;
+
+use Centreon\Domain\Repository\Interfaces\DataStorageEngineInterface;
+
+class DataStorageObserver implements DataStorageEngineInterface
+{
+    /** @var list<DataStorageEngineInterface> */
+    private array $engines = [];
+
+    /**
+     * @param DataStorageEngineInterface[] $dataStorageEngines
+     */
+    public function __construct(iterable $dataStorageEngines)
+    {
+        $this->engines = $dataStorageEngines instanceof \Traversable
+            ? iterator_to_array($dataStorageEngines)
+            : $dataStorageEngines;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function rollbackTransaction(): bool
+    {
+        $status = true;
+        foreach ($this->engines as $engine) {
+            $status = $status && $engine->rollbackTransaction();
+        }
+        return $status;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function startTransaction(): bool
+    {
+        $status = true;
+        foreach ($this->engines as $engine) {
+            $status = $status && $engine->startTransaction();
+        }
+        return $status;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function commitTransaction(): bool
+    {
+        $status = true;
+        foreach ($this->engines as $engine) {
+            $status = $status && $engine->commitTransaction();
```

Comment:
⚠️ Maybe a problem here because the second part after the `&&` won't be evaluated if `$status` is already `false`.
```suggestion
            $commited = $engine->commitTransaction();
            $status = $status && $commited;
```